### PR TITLE
Added new DaC interface

### PR DESCRIPTION
**Pull request by Drioastorga**
Added new divide_and_conquer interface that recieves the divider function and the stop condition as different arguments.

### CODE REVIEW

File: include/divideconquer.h [hunk 1]
```diff
@@ -63,6 +63,38 @@ auto divide_conquer(
...
+\param solver_op Solver operation.
+\param combiner_op Combiner operation.
+*/
+template <typename Execution, typename Input,
+          typename Divider,typename Predicate, typename Solver, typename Combiner>
+auto divide_conquer(
+    const Execution & ex,
+    Input && input,
+    Divider && divider_op,
+    Predicate && condition_op,
```
Comment (jdgarciauc3m) [inline include/divideconquer.h]: Undocumented!

In any case I suggest to rename condition_op to predicate_op

File: include/dyn/dynamic_execution.h [hunk 2]
```diff
@@ -142,6 +142,27 @@ class dynamic_execution {
...
+  \param ex Sequential execution policy object.
+  \param input Input problem to be solved.
+  \param divider_op Divider operation.
+  \param solver_op Solver operation.
+  \param combine_op Combiner operation.
+  */
+  template <typename Input, typename Divider, typename Predicate, typename Solver, typename Combiner>
+  auto divide_conquer(Input && input,
+                      Divider && divide_op,
+                      Predicate && condition_op,
```
Comment (jdgarciauc3m) [inline include/dyn/dynamic_execution.h]: Rename `divide_op` to `predicate_op` and document it.

File: include/dyn/dynamic_execution.h [hunk 3]
```diff
@@ -318,6 +339,22 @@ auto dynamic_execution::divide_conquer(
       std::forward<Combiner>(combine_op));
 }
 
+
```
Comment (jdgarciauc3m) [inline include/dyn/dynamic_execution.h:344]: Undocumented

File: include/dyn/dynamic_execution.h [hunk 4]
```diff
@@ -318,6 +339,22 @@ auto dynamic_execution::divide_conquer(
...
+template <typename Input, typename Divider,typename Predicate, typename Solver, typename Combiner>
+auto dynamic_execution::divide_conquer(
+    Input && input,
+    Divider && divide_op,
+    Predicate && condition_op,
```
Comment (jdgarciauc3m) [inline include/dyn/dynamic_execution.h]: Rename `divide_op` to `predicate_op` and document it.

File: include/native/parallel_execution_native.h [hunk 5]
```diff
@@ -331,6 +331,27 @@ class parallel_execution_native {
...
+  \param ex Sequential execution policy object.
+  \param input Input problem to be solved.
+  \param divider_op Divider operation.
+  \param solver_op Solver operation.
+  \param combine_op Combiner operation.
+  */
+  template <typename Input, typename Divider, typename Predicate, typename Solver, typename Combiner>
+  auto divide_conquer(Input && input,
+                      Divider && divide_op,
+                      Predicate && condition_op,
```
Comment (jdgarciauc3m) [inline include/native/parallel_execution_native.h]: Rename `divide_op` to `predicate_op` and document it.

File: include/native/parallel_execution_native.h [hunk 6]
```diff
@@ -351,6 +372,15 @@ class parallel_execution_native {
                       Combiner && combine_op,
                       std::atomic<int> & num_threads) const; 
 
+ template <typename Input, typename Divider,typename Predicate, typename Solver, typename Combiner>
+  auto divide_conquer(Input && input,
+                      Divider && divide_op,
+                      Predicate && condition_op,
```
Comment (jdgarciauc3m) [inline include/native/parallel_execution_native.h]: Rename `divide_op` to `predicate_op` and document it.

File: include/native/parallel_execution_native.h [hunk 7]
```diff
@@ -735,6 +765,23 @@ auto parallel_execution_native::divide_conquer(
         num_threads);
 }
 
+
+template <typename Input, typename Divider,typename Predicate, typename Solver, typename Combiner>
+auto parallel_execution_native::divide_conquer(
+    Input && problem,
+    Divider && divide_op,
+    Predicate && condition_op,
```
Comment (jdgarciauc3m) [inline include/native/parallel_execution_native.h]: Rename `divide_op` to `predicate_op` and document it.

File: include/native/parallel_execution_native.h [hunk 8]
```diff
@@ -816,6 +863,62 @@ auto parallel_execution_native::divide_conquer(
       std::forward<subresult_type>(subresult), std::forward<Combiner>(combine_op));
 }
 
+template <typename Input, typename Divider,typename Predicate, typename Solver, typename Combiner>
```
Comment (jdgarciauc3m) [inline include/native/parallel_execution_native.h:868]: Document!

File: include/native/parallel_execution_native.h [hunk 9]
```diff
@@ -816,6 +863,62 @@ auto parallel_execution_native::divide_conquer(
...
+    Predicate && condition_op,
+    Solver && solve_op,
+    Combiner && combine_op,
+    std::atomic<int> & num_threads) const
+{
+  constexpr sequential_execution seq;
+  if (num_threads.load() <=0) {
+    return seq.divide_conquer(std::forward<Input>(input),
+        std::forward<Divider>(divide_op),
+        std::forward<Predicate>(condition_op),
```
Comment (jdgarciauc3m) [inline include/native/parallel_execution_native.h]: Rename `divide_op` to `predicate_op`

File: include/omp/parallel_execution_omp.h [hunk 10]
```diff
@@ -234,6 +234,26 @@ class parallel_execution_omp {
...
+  \param ex Sequential execution policy object.
+  \param input Input problem to be solved.
+  \param divider_op Divider operation.
+  \param solver_op Solver operation.
+  \param combine_op Combiner operation.
+  */
+  template <typename Input, typename Divider, typename Predicate, typename Solver, typename Combiner>
+  auto divide_conquer(Input && input,
+                      Divider && divide_op,
+                      Predicate && condition_op,
```
Comment (jdgarciauc3m) [inline include/omp/parallel_execution_omp.h]: Rename `divide_op` to `predicate_op` and document it.

File: include/omp/parallel_execution_omp.h [hunk 11]
```diff
@@ -254,6 +274,15 @@ class parallel_execution_omp {
                       Combiner && combine_op,
                       std::atomic<int> & num_threads) const; 
 
+  template <typename Input, typename Divider, typename Predicate, typename Solver, typename Combiner>
+  auto divide_conquer(Input && input,
+                      Divider && divide_op,
+                      Predicate && condition_op,
+                      Solver && solve_op,
+                      Combiner && combine_op,
```
Comment (jdgarciauc3m) [inline include/omp/parallel_execution_omp.h]: Rename `divide_op` to `predicate_op` and document it.

File: include/omp/parallel_execution_omp.h [hunk 12]
```diff
@@ -644,6 +673,24 @@ void parallel_execution_omp::stencil(
   }
 }
 
+template <typename Input, typename Divider,typename Predicate, typename Solver, typename Combiner>
+auto parallel_execution_omp::divide_conquer(
+    Input && input,
+    Divider && divide_op,
+    Predicate && condition_op,
```
Comment (jdgarciauc3m) [inline include/omp/parallel_execution_omp.h]: Rename `divide_op` to `predicate_op` and document it.

File: include/omp/parallel_execution_omp.h [hunk 13]
```diff
@@ -690,6 +737,85 @@ void parallel_execution_omp::pipeline(
 }
 
 // PRIVATE MEMBERS
+template <typename Input, typename Divider,typename Predicate, typename Solver, typename Combiner>
+auto parallel_execution_omp::divide_conquer(
+    Input && input,
+    Divider && divide_op,
+    Predicate && condition_op,
```
Comment (jdgarciauc3m) [inline include/omp/parallel_execution_omp.h]: Rename `divide_op` to `predicate_op` and document it.

File: include/seq/sequential_execution.h [hunk 14]
```diff
@@ -174,6 +174,26 @@ class sequential_execution {
...
+  \param ex Sequential execution policy object.
+  \param input Input problem to be solved.
+  \param divider_op Divider operation.
+  \param solver_op Solver operation.
+  \param combine_op Combiner operation.
+  */
+  template <typename Input, typename Divider,typename Predicate, typename Solver, typename Combiner>
+  auto divide_conquer(Input && input,
+                      Divider && divide_op,
+                      Predicate && condition_op,
```
Comment (jdgarciauc3m) [inline include/seq/sequential_execution.h]: Rename `divide_op` to `predicate_op` and document it.

File: include/seq/sequential_execution.h [hunk 15]
```diff
@@ -437,6 +457,32 @@ constexpr void sequential_execution::stencil(
   }
 }
 
+
+template <typename Input, typename Divider, typename Predicate, typename Solver, typename Combiner>
+auto sequential_execution::divide_conquer(
+    Input && input,
+    Divider && divide_op,
+    Predicate && condition_op,
```
Comment (jdgarciauc3m) [inline include/seq/sequential_execution.h]: Rename `divide_op` to `predicate_op` and document it.

File: include/tbb/parallel_execution_tbb.h [hunk 16]
```diff
@@ -215,6 +215,26 @@ class parallel_execution_tbb {
...
+  \param ex Sequential execution policy object.
+  \param input Input problem to be solved.
+  \param divider_op Divider operation.
+  \param solver_op Solver operation.
+  \param combine_op Combiner operation.
+  */
+  template <typename Input, typename Divider,typename Predicate, typename Solver, typename Combiner>
+  auto divide_conquer(Input && input,
+                      Divider && divide_op,
+                      Predicate && condition_op,
```
Comment (jdgarciauc3m) [inline include/tbb/parallel_execution_tbb.h]: Rename `divide_op` to `predicate_op` and document it.

File: include/tbb/parallel_execution_tbb.h [hunk 17]
```diff
@@ -235,6 +255,14 @@ class parallel_execution_tbb {
                       Combiner && combine_op,
                       std::atomic<int> & num_threads) const; 
 
+  template <typename Input, typename Divider, typename Predicate, typename Solver, typename Combiner>
+  auto divide_conquer(Input && input,
+                      Divider && divide_op,
+                      Predicate && condition_op,
+                      Solver && solve_op,
+                      Combiner && combine_op,
```
Comment (jdgarciauc3m) [inline include/tbb/parallel_execution_tbb.h]: Rename `divide_op` to `predicate_op` and document it.

File: include/tbb/parallel_execution_tbb.h [hunk 18]
```diff
@@ -664,6 +708,62 @@ auto parallel_execution_tbb::divide_conquer(
       std::forward<subresult_type>(out), std::forward<Combiner>(combine_op));
 }
 
+template <typename Input, typename Divider, typename Predicate, typename Solver, typename Combiner>
+auto parallel_execution_tbb::divide_conquer(
+    Input && input,
+    Divider && divide_op,
+    Predicate && condition_op,
```
Comment (jdgarciauc3m) [inline include/tbb/parallel_execution_tbb.h]: Rename `divide_op` to `predicate_op` and document it.

**Review by jdgarciauc3m — CHANGES_REQUESTED** [review]
Address changes.

**Comment by Drioastorga** [comment]
Renamed condition_op to predicate_op and updated documentation.

**Review by jdgarciauc3m — APPROVED** [review]
OK. Done.

**Pull request merged**